### PR TITLE
feat: reaction picker SQSERVICES-1928

### PR DESCRIPTION
**Pull request by MarcinR**
<table>
<td>
  <a href="https://wearezeta.atlassian.net/browse/SQSERVICES-1928" title="SQSERVICES-1928" target="_blank">TaskSQSERVICES-1928</a>  create emoji picker view
  </td></table>
  <br />
 

Review only after merging https://github.com/wireapp/wire-ios-mono/pull/218
Create picker for all reactions
----
#### PR Submission Checklist for internal contributors

- The **PR Title**
  - [ ] conforms to the style of semantic commits messages¹ supported in Wire's Github Workflow²
  - [ ] contains a reference JIRA issue number like `SQPIT-764`
  - [ ] answers the question: _If merged, this PR will: ..._ ³

- The **PR Description**
  - [ ] is free of optional paragraphs and you have filled the relevant parts to the best of your ability
----

### CODE REVIEW

**Comment by KaterinaWire** [comment]
@MarcinR please add a description to this PR.

**Comment by KaterinaWire** [comment]
@MarcinR You can change the base branch, so we can review only this PR

File: wire-ios/Wire-iOS Tests/ConversationMessageCell/CompleteReactionPickerViewControllerTests.swift [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+
+    func testReactionPicker_scrolledToMiddle() {
+        // GIVEN WHEN
+        sut.view.subviews.forEach {
+            if let collectionView = $0 as? UICollectionView {
+                collectionView.scrollToItem(at: IndexPath(item: 0, section: 4),
+                                            at: .top,
+                                            animated: false)
+            }
+        }
```
Comment (netbe) [inline wire-ios/Wire-iOS Tests/ConversationMessageCell/CompleteReactionPickerViewControllerTests.swift]: create a function to gather common logic with testReactionPicker_scrolledToBottom.

File: wire-ios/Wire-iOS/Sources/UserInterface/Conversation/Content/MessageActions/CompleteReactionPicker/CompleteReactionPickerViewController.swift [hunk 2]
```diff
@@ -0,0 +1,161 @@
...
+        searchBar.backgroundImage = UIImage()
+        searchBar.searchTextField.backgroundColor = SemanticColors.View.backgroundDefaultWhite
+        searchBar.placeholder = Strings.search
+        view.addSubview(searchBar)
+        view.backgroundColor = SemanticColors.View.backgroundDefault
+        view.addSubview(collectionView)
+    }
+
+    private func createConstraints() {
+        guard let sectionViewControllerView = sectionViewController.view else { return }
```
Comment (netbe) [inline wire-ios/Wire-iOS/Sources/UserInterface/Conversation/Content/MessageActions/CompleteReactionPicker/CompleteReactionPickerViewController.swift:82]: why do we need this guard ? view is not optional. Did you mean?
```suggestion
        guard let sectionViewControllerView = sectionViewController. viewIfLoaded else { return }
```

File: wire-ios/Wire-iOS/Sources/UserInterface/Conversation/Content/MessageActions/CompleteReactionPicker/CompleteReactionPickerViewController.swift [hunk 3]
```diff
@@ -0,0 +1,161 @@
...
+    }
+
+    func cellForEmoji(_ emoji: Emoji, indexPath: IndexPath) -> UICollectionViewCell {
+        let cell = self.collectionView.dequeueReusableCell(withReuseIdentifier: EmojiCollectionViewCell.zm_reuseIdentifier, for: indexPath) as! EmojiCollectionViewCell
+        cell.titleLabel.text = emoji
+        return cell
+    }
+
+    func updateSectionSelection() {
+        DispatchQueue.main.async {
```
Comment (netbe) [inline wire-ios/Wire-iOS/Sources/UserInterface/Conversation/Content/MessageActions/CompleteReactionPicker/CompleteReactionPickerViewController.swift]: why do we need this dispatch?

File: wire-ios/Wire-iOS/Sources/UserInterface/Conversation/Content/MessageActions/CompleteReactionPicker/ReactionCategoryButton.swift [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program. If not, see http://www.gnu.org/licenses/.
+//
+
+import UIKit
+
+class ReactionCategoryButton: UIButton {
```
Comment (netbe) [inline wire-ios/Wire-iOS/Sources/UserInterface/Conversation/Content/MessageActions/CompleteReactionPicker/ReactionCategoryButton.swift]:
```suggestion
final class ReactionCategoryButton: UIButton {
```

File: wire-ios/Wire-iOS/Sources/UserInterface/Conversation/Content/MessageActions/CompleteReactionPicker/CompleteReactionPickerViewController.swift [hunk 5]
```diff
@@ -0,0 +1,161 @@
...
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program. If not, see http://www.gnu.org/licenses/.
+//
+
+import UIKit
+
+class CompleteReactionPickerViewController: UIViewController {
```
Comment (netbe) [inline wire-ios/Wire-iOS/Sources/UserInterface/Conversation/Content/MessageActions/CompleteReactionPicker/CompleteReactionPickerViewController.swift]:
```suggestion
final class CompleteReactionPickerViewController: UIViewController {
```

File: wire-ios/Wire-iOS/Sources/UserInterface/Conversation/Content/MessageActions/CompleteReactionPicker/ReactionSectionViewController.swift [hunk 6]
```diff
@@ -0,0 +1,148 @@
...
+    @objc private func didPan(_ recognizer: UIPanGestureRecognizer) {
+        switch recognizer.state {
+        case .possible: break
+        case .began:
+            ignoreSelectionUpdates = true
+            fallthrough
+        case .changed:
+            let location = recognizer.location(in: view)
+            guard let button = sectionButtons.filter({ $0.frame.contains(location) }).first else { return }
+            guard let type = typesByButton[button] else { return }
```
Comment (netbe) [inline wire-ios/Wire-iOS/Sources/UserInterface/Conversation/Content/MessageActions/CompleteReactionPicker/ReactionSectionViewController.swift]:
```suggestion
            guard let button = sectionButtons.filter({ $0.frame.contains(location) }).first,
                       let type = typesByButton[button]
            else { return }
```

File: wire-ios/Wire-iOS/Sources/UserInterface/Conversation/Content/MessageActions/CompleteReactionPicker/ReactionSectionViewController.swift [hunk 7]
```diff
@@ -0,0 +1,148 @@
...
+    }
+
+    override func viewDidLayoutSubviews() {
+        super.viewDidLayoutSubviews()
+        sectionButtons.forEach {
+            $0.removeFromSuperview()
+            view.addSubview($0)
+        }
+
+        createConstraints()
```
Comment (netbe) [inline wire-ios/Wire-iOS/Sources/UserInterface/Conversation/Content/MessageActions/CompleteReactionPicker/ReactionSectionViewController.swift]: I don't really get why this is done here in viewDidLayoutSubviews

File: wire-ios/Wire-iOS/Sources/UserInterface/Conversation/InputBar/Emoji/EmojiKeyboardViewController.swift [hunk 8]
```diff
@@ -129,7 +129,7 @@ final class EmojiKeyboardViewController: UIViewController {
 
 extension EmojiKeyboardViewController: EmojiSectionViewControllerDelegate {
 
-    func sectionViewController(_ viewController: EmojiSectionViewController, didSelect type: EmojiSectionType, scrolling: Bool) {
+    func sectionViewController(_ viewController: UIViewController, didSelect type: EmojiSectionType, scrolling: Bool) {
```
Comment (netbe) [inline wire-ios/Wire-iOS/Sources/UserInterface/Conversation/InputBar/Emoji/EmojiKeyboardViewController.swift]: viewController is unused so you don't have to define the type here

Comment (netbe) [reply wire-ios/Wire-iOS/Sources/UserInterface/Conversation/InputBar/Emoji/EmojiKeyboardViewController.swift]: or even remove the parameter

File: wire-ios/Wire-iOS/Sources/UserInterface/Conversation/Content/MessageActions/CompleteReactionPicker/CompleteReactionPickerViewController.swift [hunk 9]
```diff
@@ -0,0 +1,161 @@
...
+            searchBar.topAnchor.constraint(equalTo: topBar.safeBottomAnchor),
+            searchBar.leadingAnchor.constraint(equalTo: view.safeLeadingAnchor, constant: 10.0),
+            searchBar.trailingAnchor.constraint(equalTo: view.safeTrailingAnchor, constant: -10.0),
+
+            sectionViewControllerView.topAnchor.constraint(equalTo: searchBar.bottomAnchor, constant: -8.0),
+            sectionViewControllerView.leadingAnchor.constraint(equalTo: view.safeLeadingAnchor, constant: 10.0),
+            sectionViewControllerView.trailingAnchor.constraint(equalTo: view.safeTrailingAnchor, constant: -10.0),
+            sectionViewControllerView.heightAnchor.constraint(equalToConstant: 44.0),
+
+
```
Comment (agisilaos) [inline wire-ios/Wire-iOS/Sources/UserInterface/Conversation/Content/MessageActions/CompleteReactionPicker/CompleteReactionPickerViewController.swift]:
```suggestion

```

File: wire-ios/Wire-iOS Tests/ConversationMessageCell/CompleteReactionPickerViewControllerTests.swift [hunk 10]
```diff
@@ -0,0 +1,64 @@
...
+
+    var sut: CompleteReactionPickerViewController!
+
+    override func setUp() {
+        super.setUp()
+        sut = CompleteReactionPickerViewController(selectedReaction: "😀")
+        sut.view.setNeedsLayout()
+        sut.view.layoutIfNeeded()
+    }
+
```
Comment (agisilaos) [inline wire-ios/Wire-iOS Tests/ConversationMessageCell/CompleteReactionPickerViewControllerTests.swift]: Let's create a tearDown method here where we set the set to nil.

```suggestion
override func tearDown() {
        sut = nil
        super.tearDown()
    }
```

File: wire-ios/Wire-iOS Tests/ConversationMessageCell/CompleteReactionPickerViewControllerTests.swift [hunk 11]
```diff
@@ -0,0 +1,64 @@
...
+    func testReactionPicker() {
+        verify(matching: sut)
+    }
+
+    func testReactionPicker_scrolledToMiddle() {
+        // GIVEN WHEN
```
Comment (agisilaos) [inline wire-ios/Wire-iOS Tests/ConversationMessageCell/CompleteReactionPickerViewControllerTests.swift]:
```suggestion
        // GIVEN & WHEN
```

File: wire-ios/Wire-iOS Tests/ConversationMessageCell/CompleteReactionPickerViewControllerTests.swift [hunk 12]
```diff
@@ -0,0 +1,64 @@
...
+        scrollToSection(4)
+
+        // THEN
+        verify(matching: sut)
+    }
+
+
+    func testReactionPicker_scrolledToBottom() {
+        // GIVEN WHEN
```
Comment (agisilaos) [inline wire-ios/Wire-iOS Tests/ConversationMessageCell/CompleteReactionPickerViewControllerTests.swift]:
```suggestion
        // GIVEN & WHEN
```

**Review by agisilaos — COMMENTED** [review]
Left some comments..

File: wire-ios/Wire-iOS/Sources/UserInterface/Conversation/Content/MessageActions/CompleteReactionPicker/CompleteReactionPickerViewController.swift (hunk 2 above)
Comment (MarcinR) [reply wire-ios/Wire-iOS/Sources/UserInterface/Conversation/Content/MessageActions/CompleteReactionPicker/CompleteReactionPickerViewController.swift:82]: image
it 's throwing error when there's no guard

File: wire-ios/Wire-iOS Tests/ConversationMessageCell/CompleteReactionPickerViewControllerTests.swift [hunk 13]
```diff
@@ -26,42 +26,42 @@ final class CompleteReactionPickerViewControllerTests: ZMSnapshotTestCase {
 
     override func setUp() {
         super.setUp()
-        sut = CompleteReactionPickerViewController()
+        sut = CompleteReactionPickerViewController(selectedReaction: "😀")
         sut.view.setNeedsLayout()
         sut.view.layoutIfNeeded()
     }
-
+override func tearDown() {
```
Comment (netbe) [inline wire-ios/Wire-iOS Tests/ConversationMessageCell/CompleteReactionPickerViewControllerTests.swift:33]: check indentation

File: wire-ios/Wire-iOS Tests/ConversationMessageCell/CompleteReactionPickerViewControllerTests.swift [hunk 14]
```diff
@@ -26,42 +26,42 @@ final class CompleteReactionPickerViewControllerTests: ZMSnapshotTestCase {
...
+        sut = nil
+        super.tearDown()
+    }
```
Comment (netbe) [inline wire-ios/Wire-iOS Tests/ConversationMessageCell/CompleteReactionPickerViewControllerTests.swift:36]:
```suggestion
    }
    
```

**Review by netbe — APPROVED** [review]
LGTM;) just look at the indentation / formating issues

**Review by agisilaos — APPROVED** [review]

**Pull request merged**